igvc_navigation/src/mapper/main.cpp

```diff
@@ -13,221 +15,281 @@
-    fromOcuGrid->header.stamp = msg->header.stamp;
-    debug_pcl_pub.publish(fromOcuGrid);
-  }
-  update = true;
+//  update = true;
 }
 
+
 int main(int argc, char **argv)
 {
-  ros::init(argc, argv, "mapper");
-  ros::NodeHandle nh;
-  ros::NodeHandle pNh("~");
-
-  std::list<ros::Subscriber> subs;
-  tf_listener = new tf::TransformListener();
-
-  double cont_start_x;
-  double cont_start_y;
-
-  if (!pNh.hasParam("topics") && !pNh.hasParam("occupancy_grid_length") && !pNh.hasParam("occupancy_grid_length") &&
-      !pNh.hasParam("occupancy_grid_resolution") && !pNh.hasParam("start_X") & !pNh.hasParam("start_Y") &&
-      !pNh.hasParam("debug"))
-  {
-    ROS_ERROR_STREAM("missing parameters; exiting");
-    return 0;
-  }
-
-  // assumes all params inputted in meters
-  pNh.getParam("topics", topics);
-  pNh.getParam("occupancy_grid_length", length_y);
-  pNh.getParam("occupancy_grid_width", width_x);
-  pNh.getParam("occupancy_grid_resolution", resolution);
-  pNh.getParam("start_X", cont_start_x);
-  pNh.getParam("start_Y", cont_start_y);
-  pNh.getParam("orientation", orientation);
-  pNh.getParam("debug", debug);
-
-  // convert from meters to grid
-  length_y = (int)std::round(length_y / resolution);
-  width_x = (int)std::round(width_x / resolution);
-  start_x = (int)std::round(cont_start_x / resolution);
-  start_y = (int)std::round(cont_start_y / resolution);
-  ROS_INFO_STREAM("cv::Mat length: " << length_y << "  width: " << width_x << "  resolution: " << resolution);
-
-  // set up tokens and get list of subscribers
-  std::istringstream iss(topics);
-  std::vector<std::string> tokens{ std::istream_iterator<std::string>(iss), std::istream_iterator<std::string>() };
-  odom_sub = nh.subscribe("/odometry/filtered", 1, odom_callback);
-
-  for (std::string topic : tokens)
-  {
-    ROS_INFO_STREAM("Mapper subscribing to " << topic);
-    subs.push_back(nh.subscribe<pcl::PointCloud<pcl::PointXYZ>>(topic, 1, boost::bind(frame_callback, _1, topic)));
-  }
-
-  published_map = new cv::Mat(length_y, width_x, CV_8UC1);
-  eigenRep = new Eigen::Map<Eigen::Matrix<unsigned char, Eigen::Dynamic, Eigen::Dynamic>>(
-      (unsigned char *)published_map, width_x, length_y);
-
-  map_pub = nh.advertise<igvc_msgs::map>("/map", 1);
-
-  if (debug)
-  {
-    debug_pub = nh.advertise<sensor_msgs::Image>("/map_debug", 1);
-    debug_pcl_pub = nh.advertise<pcl::PointCloud<pcl::PointXYZRGB>>("/map_debug_pcl", 1);
-  }
-
-  ros::spin();
+    ros::init(argc, argv, "mapper");
+    ros::NodeHandle nh;
+    ros::NodeHandle pNh("~");
+    std::string topics;
+
+    std::list<ros::Subscriber> subs;
+    tf_listener = std::unique_ptr<tf::TransformListener>(new tf::TransformListener());
+
+    double cont_start_x;
+    double cont_start_y;
+    double decay_period;
+    int cont_occupancy_grid_threshold;
+
+    if (!(pNh.hasParam("topics") && pNh.hasParam("occupancy_grid_width") && pNh.hasParam("occupancy_grid_length") &&
```

Comment:
this would be the proper place to implement `NodeUtils::param()` or `NodeUtils::getParam()`

Reply:
It's a separate PR, so I intended for the change to happen during merging, which is why I didn't use the `igvc::getParam()` and `igvc::param()` here.